### ISSUE TITLE
The ca_extend::extend_ca_cert plan fails if the primary certificate has expired

### ISSUE THREAD

**Issue by Sharpie** [puppetlabs/ca_extend#2]
## Describe the Bug

If the expiration date for the primary certificate in a Puppet Installation has already passed, the `ca_extend::extend_ca_cert` plan will successfully extend the CA certificate but then fail when `puppet infrastructure configure --no-recover` is run to move the new CA certificate into place.

## Expected Behavior

The first step of recovering from an expired CA completes successfully, even if the PKI infrastructure of a Puppet installation has completely expired.

## Steps to Reproduce
- Acquire a virtual machine running CentOS 7.

- Set the system clock back in time 5 years:

```bash
timedatectl set-ntp false
timedatectl set-time "$(date --date='5 years ago' +'%Y-%m-%d %H:%M:%S')"
```

- Install PE 2018.1.18. Run `puppet agent -t` twice to verify everything works.

- Re-enable NTP with `timedatectl set-ntp true`. `puppet agent -t` should now fail as all certificates have expired.

- Run the `ca_extend::extend_ca_cert` plan. It will fail while trying to run `puppet infrastructure configure`:

```
# bolt --version
2.44.0

# bolt puppetfile show_modules
# bolt puppetfile show-modules
...
/root/Boltdir/modules
  puppetlabs/ca_extend   1.2.1
  puppetlabs/stdlib      6.6.0
...

# bolt plan run ca_extend::extend_ca_cert --targets "local://$(puppet config print certname)"
Starting: plan ca_extend::extend_ca_cert
Starting: install puppet and gather facts on local://20198-primary.p9.test
The local transport will default to using Bolt's Ruby interpreter and setting the 'puppet-agent' feature in Bolt 3.0. Enable or disable these defaults by setting 'bundled-ruby' in the local transport config. [ID: local_default_config]
Finished: install puppet and gather facts with 0 failures in 9.79 sec
Starting: task facts on local://20198-primary.p9.test
Finished: task facts with 0 failures in 11.05 sec
INFO: Stopping Puppet services on local://20198-primary.p9.test
Starting: task service::linux on local://20198-primary.p9.test
Finished: task service::linux with 0 failures in 0.13 sec
Starting: task service::linux on local://20198-primary.p9.test
Finished: task service::linux with 0 failures in 1.22 sec
Starting: task service::linux on local://20198-primary.p9.test
Finished: task service::linux with 0 failures in 0.16 sec
INFO: Extending CA certificate on local://20198-primary.p9.test
Starting: task ca_extend::extend_ca_cert on local://20198-primary.p9.test
Finished: task ca_extend::extend_ca_cert with 0 failures in 2.42 sec
INFO: Configuring local://20198-primary.p9.test to use the extended CA certificate
Starting: task ca_extend::configure_master on local://20198-primary.p9.test
Finished: task ca_extend::configure_master with 1 failure in 82.77 sec
Finished: plan ca_extend::extend_ca_cert in 1 min, 48 sec
Failed on local://20198-primary.p9.test:
  The task failed with exit code 1
  2021-02-05T01:15:53.722+00:00 - [Notice]: Compiled catalog for 20198-primary.p9.test in environment enterprise in 5.25 seconds
  2021-02-05T01:15:59.716+00:00 - [Notice]: /Stage[main]/Puppet_enterprise::Profile::Controller/File[/etc/puppetlabs/client-tools/ssl/certs/ca.pem]/content: content changed '{md5}23cbdcd36f1cec6b03e0a7e830d6be9c' to '{md5}d1cd9af62234c6c5e48636aca1b7811d'
  2021-02-05T01:16:00.843+00:00 - [Notice]: /Stage[main]/Pe_install::Install/Service[puppet]/ensure: ensure changed 'stopped' to 'running'
  2021-02-05T01:16:07.264+00:00 - [Notice]: /Stage[main]/Pe_postgresql::Server::Service/Service[postgresqld]/ensure: ensure changed 'stopped' to 'running'
  2021-02-05T01:16:44.655+00:00 - [Notice]: /Stage[main]/Puppet_enterprise::Master::Puppetserver/Puppet_enterprise::Trapperkeeper::Pe_service[puppetserver]/Service[pe-puppetserver]/ensure: ensure changed 'stopped' to 'running'
  { "status": "error", "message": "Error running 'puppet infrastructure configure'", "stderr": "Error: Failed to initialize SSL: The certificate 'CN=20198-primary.p9.test' has expired, verify time is synchronized Error: Run `puppet agent -t` Error: Failed to initialize SSL: The certificate 'CN=20198-primary.p9.test' has expired, verify time is synchronized Error: Run `puppet agent -t` 2021-02-05T01:16:48.076+00:00 - [Error]: Failed to initialize SSL: The certificate 'CN=20198-primary.p9.test' has expired, verify time is synchronized 2021-02-05T01:16:48.076+00:00 - [Error]: Run `puppet agent -t` 2021-02-05T01:16:48.077+00:00 - [Error]: Could not prefetch pe_node_group provider 'ruby': The certificate 'CN=20198-primary.p9.test' has expired, verify time is synchronized 2021-02-05T01:16:48.251+00:00 - [Error]: Failed to apply catalog: The certificate 'CN=20198-primary.p9.test' has expired, verify time is synchronized " }
Failed on 1 target: local://20198-primary.p9.test
```

**Comment by Sharpie** [comment]
Note that the setup above introduces an additional wrinkle in that the Certificate Revocation List (CRL) on the primary node is also expired. This would be very unusual to find in a real-world deployment as there has almost always been some certificate revocation activity that would have advanced the CRL expiration on the primary. However, the first step of CA extension should be fortified to cover this issue as well.

(NOTE: An expired CRL may be common on agents as most Puppet versions never refresh the CRL on the agent side. This should be considered a different issue as it affects steps further along in the extension process.)

**Comment by Sharpie** [comment]
Closing as the failure from expired certificates was fixed by puppetlabs/ca_extend#10 which added the `regen_primary_cert` option.

Opened puppetlabs/ca_extend#15 to track the rarer remaining failure from an expired CRL.
